kernel: make kmod-*-core selected by dependent modules

mans0n opened the pull request:
Currently kmod-X-* will not get into images unless kmod-X-core is added to
DEVICE_PACKAGES as well. By changing the dependencies from "depends on" to
"select", we do not have the issue anymore.

Furthermore, we can remove most occurrences of the package from DEVICE_PACKAGES
and similar variables, as it is now pulled by dependent modules.

mans0n commented:
`git grep "kmod-[^ ]*-core"` shows the list of kmod-*-core packages:

- kmod-ata-core
- kmod-hwmon-core
- kmod-i2c-core
- kmod-input-core
- kmod-pcmcia-core
- kmod-scsi-core
- kmod-sound-core
- kmod-sound-hda-core
- kmod-video-core
- kmod-video-gspca-core

I've dealt with ata, hwmon, i2c packages here. Maybe we can also add sound and video packages...

And maybe we'd better split the commits into kernel and treewide, as we did in #2823 for check-size?

adrianschmutzler commented:
I'd start with the packages here and if they are merged and nobody cries, you are welcome to deal with the rest :-)

adrianschmutzler commented:
Do you have any idea why the dependencies were created without "+" in the first place? To me it looks like it's just ancient code and nobody cared, but I stopped looking when I was at year 2010 ...

mans0n commented:
My guess is that devs were following the Kconfig in the kernel, and the "+" for kmod-usb-core was rathter a mistake. I think adding "+" really simplify dependencies for OpenWrt, but perhaps we'd better remove "+" for kmod-usb-core if we follow the principle...

adrianschmutzler commented:
I see it from a pragmatic point of view: Selecting a module should ensure it's working, and select everything required for that.

Will kmod-ata-ahci work without kmod-ata-core? If the answer is no, selecting kmod-ata-work should also select kmod-ata-core.

mans0n commented:
Yeah, I agree with you. (Anyway, opkg does not care about it. Only build scripts care.) But then why do we need "depends on" at the first place? Should we add "+" to all DEPENDS (for packages that do not depend on CONFIG)? That was the question I asked in the mailing list.

adrianschmutzler commented:
> But then why do we need "depends on" at the first place?

From the wiki, having a package without "+" actually seems to have a different purpose:
https://openwrt.org/docs/guide-developer/dependencies#dependency_types

However, this logic seems to be just not applicable when a package is selected from the code and not by the user.

mans0n commented:
Thanks for the page, I did not know docs other then the general package guide.

> However, this logic seems to be just not applicable when a package is selected from the code and not by the user.

You have a point. For menuconfig "depends on" helps to trim down package lists, but for build scripts it does not really help anything...

adrianschmutzler commented:
I've pushed the ata commit for now, since I just had the time for reviewing and testing that one.

mans0n commented:
To make the changes easier to review at a glance, I removed the reformatting here. It will be done by a separate commit later.

adrianschmutzler commented:
I've just pushed the rest. I'm not closing this in case you want to add similar patches for the other core packages (which I'd encourage). If you decide otherwise, just close the PR yourself.

Note that I've dropped changes on ar71xx target, as this is in a semi-frozen state only used for backporting, and I do not see a point in applying cosmetic changes to it which won't be backported anyway, and thus would only complicate backporting other fixes.

While reviewing i2c changes, I realized that kmod-i2c-mux also is frequently implemented as a non-selecting depend. I'd prefer to also have that changed to +kmod-i2c-mux, go ahead if you want to do that as well, otherwise I'd pick that topic up later.

Thanks for your work on tidying up so far!

mans0n commented:
> Note that I've dropped changes on ar71xx target, as this is in a semi-frozen state only used for backporting, and I do not see a point in applying cosmetic changes to it which won't be backported anyway, and thus would only complicate backporting other fixes.

Okay, I won't include ar71xx in next patches.

> While reviewing i2c changes, I realized that kmod-i2c-mux also is frequently implemented as a non-selecting depend. I'd prefer to also have that changed to +kmod-i2c-mux, go ahead if you want to do that as well, otherwise I'd pick that topic up later.

That's also what I've noticed, but I could not decide whether I should deal with non-core packages. Now that you said it, I'll take care of it as well.

mans0n committed:
kernel: make kmod-i2c-mux selected by dependent modules

Currently kmod-i2c-mux-* will not get into images unless kmod-i2c-mux is added
to DEVICE_PACKAGES as well. By changing the dependencies from "depends on" to
"select", we do not have the issue anymore.

Furthermore, we can remove most occurrences of the package from DEVICE_PACKAGES
and similar variables, as it is now pulled by dependent modules such as:
- kmod-i2c-mux-pca954x

Signed-off-by: Sungbo Eo <mans0n@gorani.run>

mans0n commented:
Added kmod-i2c-mux patch. I decided not to deal with other packages here.

adrianschmutzler commented:
Less impact than I thought.

Thanks again, merged.
